File: DataStructure/java/DoublyLinkedList.java
package datastructure;

public class DoublyLinkedList {
    private class Node {
        private int val;
        private Node next;
        private Node prev;

        Node(final int val) {
            this.val = val;
            this.next = null;
            this.prev = null;
        }
    }

    private Node head;
    private Node tail;

    public DoublyLinkedList() {
        this.head = new Node(0);
        this.tail = new Node(0);

        head.next = tail;
        tail.prev = head;
    }

    public void pushBack(int num) {
        Node node = new Node(num);

        node.prev = tail.prev;
        node.next = tail;

        tail.prev.next = node;
        tail.prev = node;
    }

    public void pushFront(int num) {
        Node node = new Node(num);

        node.prev = head;
        node.next = head.next;

        head.next.prev = node;
        head.next = node;
    }

    public void pushSequence(int num) {
        Node temp = findNextNode(num);
        Node node = new Node(num);

        node.prev = temp.prev;
        node.next = temp;

        if (temp.prev != null) temp.prev.next = node;
        temp.prev = node;
    }

    public void deleteNode(final int num) {
        Node node = findNode(num);
        if (node == null) {
            throw new IllegalArgumentException("해당하는 원소가 없습니다.");
        }

        if (node.prev != null) node.prev.next = node.next;
        if (node.next != null) node.next.prev = node.prev;
    }

    private Node findNode(final int num) {
        Node node = head.next;
        while (node.next != null) {
            if (node.val == num) {
                return node;
            }
            node = node.next;
        }
        return null;
    }

    private Node findNextNode(final int num) {
        Node node = head;
        while (node.next != null) {
            if (node.next.val > num) {
                return node.next;
            }
            node = node.next;
        }
        return node;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        Node node = head.next;
        while (node.next != null) {
            sb.append(node.val).append(' ');
            node = node.next;
        }
        return sb.toString();
    }
}

File: DataStructure/c++/DoublyLinkedList.cpp
#include<iostream>

using namespace std;

#define MAX_NODE 10000

struct Node {
	int val;
	Node *next;
	Node *prev;
};

Node pool[MAX_NODE];
int nCnt;

Node *newNode(const int val) {
	Node *n = &pool[nCnt++];
	n->val = val;
	n->prev = 0;
	n->next = 0;
	return n;
}

struct LinkedList {
	Node *head;
	Node *tail;

	void init() {
		head = newNode(-987654321);
		tail = newNode(-987654321);
		head->next = tail;
		tail->prev = head;
	}

	void push_back(const int val) {
		Node *node = newNode(val);
		
		node->prev = tail->prev;
		node->next = tail;

		tail->prev->next = node;
		tail->prev = node;
	}

	void push_front(const int val) {
		Node *node = newNode(val);

		node->next = head->next;
		node->prev = head;

		head->next->prev = node;
		head->next = node;
	}

	Node *findSeq(const int val) {
		Node *itr = head;
		while (itr->next) {
			if (val <= itr->next->val)
				return itr->next;

			itr = itr->next;
		}
		return itr;
	}

	void push_sequence(const int val) {
		Node *node = newNode(val);
		Node *findNode = findSeq(val);

		node->prev = findNode->prev;
		node->next = findNode;

		findNode->prev->next = node;
		findNode->prev = node;
	}

	void printList() {
		Node *itr = head->next;
		while (itr->next) {
			cout << itr->val << ' ';
			itr = itr->next;
		}
		cout << '\n';
	}
};

LinkedList list;

int main() {
	list.init();
	list.push_back(2);
	list.push_back(5);
	list.push_back(8);
	list.push_back(1);
	list.printList();

	list.init();
	list.printList();

	list.push_sequence(3);
	list.push_sequence(1);
	list.push_sequence(8);
	list.push_sequence(2);
	list.printList();
}